extensions/amp-a4a/0.1/refresh-manager.js

```diff
@@ -0,0 +1,366 @@
+    this.refreshInterval_ = this.getPublisherSpecifiedRefreshInterval_();
+
+    /** @const @private {!RefreshConfig} */
+    this.config_ = this.convertConfiguration_(config);
+
+    /** @const @private {!../../../src/service/timer-impl.Timer} */
+    this.timer_ = Services.timerFor(this.win_);
+
+    /** @private {?(number|string)} */
+    this.refreshTimeoutId_ = null;
+
+    /** @private {?(number|string)} */
+    this.visibilityTimeoutId_ = null;
+
+    /** @private {boolean} */
+    this.isRefreshable_ = isExperimentOn(this.win_, 'amp-ad-refresh') &&
+        // The network has opted-in.
+        !!(this.config_ &&
+        // The publisher has enabled refresh on this slot.
+        (this.refreshInterval_ || this.refreshInterval_ == ''));
+
+    if (this.isRefreshable_) {
+      const managerId = String(refreshManagerIdCounter++);
+      this.element_.setAttribute(DATA_MANAGER_ID_NAME, managerId);
+      managers[managerId] = this;
+      this.initiateRefreshCycle();
+    }
+  }
+
+  /**
+   * Returns an IntersectionObserver configured to the given threshold, creating
+   * one if one does not yet exist.
+   *
+   * @param {(string|number)} threshold
+   * @return {(!IntersectionObserver|!IntersectionObserverPolyfill)}
+   */
+  getIntersectionObserverWithThreshold_(threshold) {
+    threshold = String(threshold);
+    return observers[threshold] ||
+        (observers[threshold] = 'IntersectionObserver' in this.win_
+         ? new this.win_['IntersectionObserver'](this.ioCallback_, {threshold})
+         : new IntersectionObserverPolyfill(this.ioCallback_, {threshold}));
+  }
+
+  /**
+   * Returns a function that will be invoked directly by the
+   * IntersectionObserver implementation. It will implement the core logic of
+   * the refresh lifecycle, including the transitions of the DFA.
+   *
+   * @param {!Array<!IntersectionObserverEntry>} entries
+   */
+  ioCallback_(entries) {
+    entries.forEach(entry => {
+      const refreshManagerId = entry.target.getAttribute(DATA_MANAGER_ID_NAME);
+      dev().assert(refreshManagerId);
+      const refreshManager = managers[refreshManagerId];
+      if (entry.target != refreshManager.element_) {
+        return;
+      }
+      switch (refreshManager.state_) {
+        case RefreshLifecycleState.INITIAL:
+          // First check if the element qualifies as "being on screen", i.e.,
+          // that at least a minimum threshold of pixels is on screen. If so,
+          // begin a timer, set for the duration of the minimum time on screen
+          // threshold. If this timer runs out without interruption, then all
+          // viewability conditions have been met, and we can begin the refresh
+          // timer.
+          if (entry.intersectionRatio >=
+              refreshManager.config_.visiblePercentageMin) {
+            refreshManager.state_ = RefreshLifecycleState.VIEW_PENDING;
+            refreshManager.visibilityTimeoutId_ = refreshManager.timer_.delay(
```

Comment:
We decided to use the IAB standard definition of "viewability". I think since this value is passed down by an implementing network, and not a publisher, I don't think it should be an issue.